File: 03541. Find Most Frequent Vowel and Consonant/solution.go
/**
 * Problem: 3541. Find Most Frequent Vowel and Consonant
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxFreqSum(s string) int {
	// Create a map to store the frequency of each rune in the string
	frequencyMap := make(map[rune]int)
	// Initialize variables to store the maximum frequency of consonants and vowels
	maxConsonantFreq, maxVowelFreq := 0, 0

	// Iterate over each character in the string and update its frequency in the map
	for _, char := range s {
		frequencyMap[char]++
	}

	// Iterate over the frequency map to find the highest frequency of vowels and consonants
	for ch, count := range frequencyMap {
		// Check if the character is a vowel
		if strings.ContainsRune("aeiou", ch) {
			// Update maxVowelFreq if the current vowel's frequency is higher
			if count > maxVowelFreq {
				maxVowelFreq = count
			}
		} else {
			// Update maxConsonantFreq if the current consonant's frequency is higher
			if count > maxConsonantFreq {
				maxConsonantFreq = count
			}
		}
	}

	// Return the sum of the highest vowel and consonant frequencies
	return maxConsonantFreq + maxVowelFreq
}

File: 03541. Find Most Frequent Vowel and Consonant/solution.js
/**
 * Problem: 3541. Find Most Frequent Vowel and Consonant
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Returns sum of highest vowel and consonant frequencies in a string
 *
 * @param {string} s - Input string
 *
 * @returns {number} - Sum of max vowel and consonant frequencies
 */
const maxFreqSum = (s) => {
  // Map to store frequency of each vowel character
  const vowelFrequency = new Map(),
    // Map to store frequency of each consonant character
    consonantFrequency = new Map()

  // Variable to track the highest frequency among vowels
  let maxVowelFreq = 0,
    // Variable to track the highest frequency among consonants
    maxConsonantFreq = 0

  // Iterate through each character in the input string
  for (const char of s) {
    // Check if the character is a vowel
    if (
      char === 'a' ||
      char === 'e' ||
      char === 'i' ||
      char === 'o' ||
      char === 'u'
    ) {
      // Get current count for this vowel, increment by 1
      const currentVowelCount = (vowelFrequency.get(char) || 0) + 1

      // Update the vowel frequency map
      vowelFrequency.set(char, currentVowelCount)

      // Update max vowel frequency if current is higher
      if (currentVowelCount > maxVowelFreq) maxVowelFreq = currentVowelCount
    } else {
      // Get current count for this consonant, increment by 1
      const currentConsonantCount = (consonantFrequency.get(char) || 0) + 1

      // Update the consonant frequency map
      consonantFrequency.set(char, currentConsonantCount)

      // Update max consonant frequency if current is higher
      if (currentConsonantCount > maxConsonantFreq)
        maxConsonantFreq = currentConsonantCount
    }
  }

  // Return the sum of the highest vowel and consonant frequencies
  return maxVowelFreq + maxConsonantFreq
}
